File: resources/code-template/node/deviceManageDeviceTwin.js
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

'use strict';

// Run 'npm install azure-iot-device-mqtt' to install the required libraries for this application
var Client = require('azure-iot-device').Client;
var Protocol = require('azure-iot-device-mqtt').Mqtt;

var connectionString = '{{deviceConnectionString}}';

// create the IoTHub client
var client = Client.fromConnectionString(connectionString, Protocol);
console.log('got client');

// connect to the hub
client.open(function(err) {
  if (err) {
    console.error('could not open IotHub client');
    process.exit(-1);
  } else {
    console.log('client opened');

    // Create device Twin
    client.getTwin(function(err, twin) {
      if (err) {
        console.error('could not get twin');
        process.exit(-1);
      } else {
        console.log('twin created');

        // set up code to handle desired property changes.
        twin.on('properties.desired', function(delta) {
            console.log('new desired properties received:');
            console.log(JSON.stringify(delta));
        });

        // create a patch to send to the hub
        var patch = {
          firmwareVersion:'1.2.2',
          weather:{
            temperature: 72,
            humidity: 17
          }
        };

         // send the patch
        twin.properties.reported.update(patch, function(err) {
          if (err) {
            console.error('unable to update twin: ' + err.toString());
            process.exit(-1);
          } else {
            console.log('twin state reported');
          }
        });
      }
    });
  }
});

File: resources/code-template/python/deviceManageDeviceTwin.py
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project root for full license information.

import random
import time
import sys

# Using the Python Device SDK for IoT Hub:
#   https://github.com/Azure/azure-iot-sdk-python
#   Run 'pip install azure-iot-device' to install the required libraries for this application

import asyncio
from azure.iot.device.aio import IoTHubDeviceClient

# The device connection string to authenticate the device with your IoT hub.
CONNECTION_STRING = "{{deviceConnectionString}}"

async def manage_device_twin_sample_run():
    try:
        # Create instance of the device client using the authentication provider
        device_client = IoTHubDeviceClient.create_from_connection_string(CONNECTION_STRING)

        # Connect the device client.
        await device_client.connect()

        print ( "Get device twin" )
        twin = await device_client.get_twin()
        print (twin)

        print ( "Updating reported properties" )
        reported_properties_patch = {
            "weather": {
                "temperature": 72,
                "humidity": 17
            },
            "firmwareVersion": "1.2.2"
        }
        await device_client.patch_twin_reported_properties(reported_properties_patch)

    except Exception as iothub_error:
        print ( "Unexpected error %s from IoTHub" % iothub_error )
    finally:
        print ( "Disconnecting client" )
        await device_client.disconnect()

if __name__ == '__main__':
    print ( "IoT Hub simulated device" )
    asyncio.run(manage_device_twin_sample_run())

    # If using Python 3.6 or below, use the following code instead of asyncio.run(manage_device_twin_sample_run()):
    # loop = asyncio.get_event_loop()
    # loop.run_until_complete(manage_device_twin_sample_run())
    # loop.close()